linkerd/retry/src/buf_body.rs

```diff
@@ -0,0 +1,689 @@
+#![allow(dead_code)]
+
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use http::HeaderMap;
+use http_body::Body;
+use linkerd_stack as svc;
+use std::{
+    collections::VecDeque,
+    io::IoSlice,
+    mem,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::Context,
+    task::Poll,
+};
+
+/// Wraps an HTTP body type and lazily buffers data as it is read from the inner
+/// body.
+///
+/// When the initial `Body` is dropped, any data in the buffer is transferred to
+/// the cloned body, if it exists.
+///
+/// The buffered data can then be used to retry the request if the original
+/// request fails.
+#[derive(Debug)]
+pub struct BufBody<B: Body + Default> {
+    inner: Inner<B>,
+}
+
+/// Data returned by `BufBody`'s `http_body::Body` implementation is either
+/// `Bytes` returned by the initial body, or a list of all `Bytes` chunks
+/// returned by the initial body (when replaying it).
+#[derive(Debug)]
+pub enum Data {
+    Initial(Bytes),
+    Replay(BufList),
+}
+
+/// Body data composed of multiple `Bytes` chunks.
+#[derive(Debug, Default)]
+pub struct BufList {
+    bufs: VecDeque<Bytes>,
+}
+
+/// A `Service`/`Proxy` that wraps an HTTP request's `Body` type in a `BufBody`,
+/// allowing it to be cloned.
+///
+// TODO(eliza): it would be nice if this could just be `MapRequest`, but Tower
+// won't let us implement `Proxy` for it (because it doesn't expose access to
+// the closure). Maybe we can change that upstream eventually...
+#[derive(Debug, Default, Clone)]
+pub struct WrapBody<S> {
+    inner: S,
+}
+
+#[derive(Debug)]
+enum Inner<B: Body + Default> {
+    Initial(InitialBody<B>),
+    Replay(ReplayBody<B>),
+}
+
+#[derive(Debug)]
+struct InitialBody<B: Body + Default> {
+    body: B,
+    bufs: BufList,
+    trailers: Option<HeaderMap>,
+    shared: Arc<Mutex<Option<BodyState<B>>>>,
+    is_completed: bool,
+}
+
+#[derive(Debug)]
+struct BodyState<B> {
+    body: Option<BufList>,
+    trailers: Option<HeaderMap>,
+    rest: Option<B>,
+    is_completed: bool,
+}
+
+#[derive(Debug)]
+enum ReplayBody<B> {
+    Waiting(Arc<Mutex<Option<BodyState<B>>>>),
+    Ready(BodyState<B>),
+    Empty,
+}
+
+// === impl BufBody ===
+
+impl<B: Body + Default> BufBody<B> {
+    /// Wraps an initial `Body` in a `BufBody`.
+    pub fn new(body: B) -> Self {
+        Self {
+            inner: Inner::Initial(InitialBody {
+                body,
+                bufs: Default::default(),
+                trailers: None,
+                shared: Arc::new(Mutex::new(None)),
+                is_completed: false,
+            }),
+        }
+    }
+
+    /// If this is the initial body, returns a `Some` with a replay body.
+    pub fn try_clone(&self) -> Option<Self> {
+        match self.inner {
+            Inner::Initial(InitialBody { ref shared, .. }) => Some(Self {
+                inner: Inner::Replay(ReplayBody::Waiting(shared.clone())),
+            }),
+            _ => None,
+        }
+    }
```

Comment:
I don't think I've spotted anything that explicitly addresses this in the above comments: does this only support a single retry for a request with a body? If so, why does this limitation exist? What's required to support an arbitrary number of retries?

Reply:
ah, hmm, the main thing is just that the `VecDeque` of buffers is moved by value into and out of the replay body, so that we don't have to clone it. if we wanted to support cloning the body multiple times, we would need to clone it so that it can be (potentially) used by a subsequent retry...

Reply:
i'm going to go ahead and change this so that the body can be cloned again